tests/Agency/StorePerformanceTest.cpp

```diff
@@ -0,0 +1,666 @@
+
+namespace arangodb {
+namespace tests {
+namespace store_performance_test {
+
+const std::array<size_t, 4> repetition_times {100, 1500, 5000, 20000};
+
+struct OperationMeasurement 
+{
+  using Clock = std::chrono::high_resolution_clock;
+  using Duration = Clock::duration;
+  using DurationIterator = std::vector<Duration>::iterator;
+  using InformValue = std::function<void(Duration)>;
+  using StatExtractor = std::function<void(DurationIterator,DurationIterator,InformValue)>;
+  using Observation = Clock::time_point;
+  using Observations = std::vector<Observation>;
+  
+  OperationMeasurement(size_t expected_count = 0u) {
+    observations_.reserve(expected_count + 2);
+    ++*this;
+  }
+
+  OperationMeasurement(Observations&& source): observations_{source} {}
+
+  ~OperationMeasurement() {
+    stop();
+  }
+
+  void stop() {
+    if (!stopped_) {
+      if (observations_.size() < 2) {
+        // we must be measuring a single long operation
+        ++*this;
+      }
+      stopped_ = true;
+    }
+  }
+
+  OperationMeasurement(OperationMeasurement const&) = delete;
+  OperationMeasurement(OperationMeasurement&) = delete;
+  OperationMeasurement(OperationMeasurement&&) = delete;
+
+  OperationMeasurement& operator++() {
+    observations_.push_back(Clock::now());
+    return *this;
+  }
+
+  void report(size_t block_size = 0) 
+  {
+    stop();
+    static std::array<std::pair<std::string_view, StatExtractor>, 5> duration_stats = {{
+      {"max", [](DurationIterator begin, DurationIterator end, InformValue inform) { inform (*std::max_element(begin,end)); }},
+      {"min", [](DurationIterator begin, DurationIterator end, InformValue inform) { inform (*std::min_element(begin,end)); }},
+      {"avg", [](DurationIterator begin, DurationIterator end, InformValue inform) { 
+        if (begin != end) { 
+          inform (std::accumulate(begin, end, OperationMeasurement::Clock::duration::zero()) / std::distance(begin, end));
+        }
+      }},
+      {"med", [](DurationIterator begin, DurationIterator end, InformValue inform) { 
+        if (begin != end) { 
+          std::vector<Duration> copy {begin, end};
+          std::sort(copy.begin(), copy.end());
+          inform (copy[copy.size() / 2]);
+        }
+      }},
+      {"max10", [](DurationIterator begin, DurationIterator end, InformValue inform) { 
+        if (begin != end) { 
+          std::vector<Duration> copy {begin, end};
+          std::sort(copy.begin(), copy.end(), [](Duration const& a, Duration const& b){ return b < a; });
+          copy.resize(std::min(10ul, copy.size()));
+          std::for_each(copy.begin(), copy.end(), inform);
+        }
+      }},
+    }};
+
+    std::vector<OperationMeasurement::Clock::duration> durations;
+    auto begin{observations_.begin()};
+    auto from {*begin};
+    auto const absolute_start {*begin};
+    size_t block_position {0};
+    std::transform(++begin, observations_.end(), std::back_inserter(durations), 
```

Comment:
This code is rather obfuscated. I had a hard time to understand it. We either need better comments or a different approach.